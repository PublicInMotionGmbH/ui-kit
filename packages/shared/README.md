# Talixo Shared

Shared data between whole UI Kit.

## How to install

Package is available as `@talixo/shared` in NPM registry, so you can use it in your project
using `npm install @talixo/shared --save` or `yarn add @talixo/shared`.

## Requirements

Your package should additionally have some extra dependencies:

- `classnames: ^2.2.5`

These packages are required by `@talixo/shared`, but you have to install them manually,
to avoid having different versions of these in your application.
