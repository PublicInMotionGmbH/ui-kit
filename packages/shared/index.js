import applyAnyClassNameModifiers from './src/applyAnyClassNameModifiers'
import applyClassNameModifiers from './src/applyClassNameModifiers'
import buildClassName from './src/buildClassName'
import prefix from './src/prefix'

export {
  applyAnyClassNameModifiers,
  applyClassNameModifiers,
  buildClassName,
  prefix
}
