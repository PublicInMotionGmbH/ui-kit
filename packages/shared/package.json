{
  "name": "@talixo/shared",
  "version": "1.0.0",
  "description": "Talixo common configuration",
  "author": "Public in Motion GmbH",
  "main": "index.js",
  "module": "index.js",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/PublicInMotionGmbH/ui-kit.git"
  },
  "bugs": {
    "url": "https://github.com/PublicInMotionGmbH/ui-kit/issues"
  },
  "devDependencies": {
    "rollup": "^0.57.1"
  },
  "dependencies": {
    "normalize.css": "^8.0.0"
  }
}
