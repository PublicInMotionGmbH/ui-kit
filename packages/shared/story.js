import { storiesOf } from '@storybook/react'
import { withInfo } from '@storybook/addon-info'

import createController from './story/createController'

/**
 * Build styles for story
 *
 * @param {object} stylesheet
 * @param {object} options
 * @returns {object}
 */
const styles = (stylesheet, options) => ({
  ...stylesheet,
  infoBody: {
    ...stylesheet.infoBody,
    boxShadow: 'none',
    border: 0,
    padding: 0
  },
  propTableHead: options.propTables && options.propTables.length > 1 ? {

  } : {
    display: 'none'
  },
  source: {
    ...stylesheet.source,
    h1: {
      ...stylesheet.source.h1,
      marginBottom: '15px'
    }
  }
})

/**
 * Create factory for stories
 *
 * @param {string} name
 * @param {object} module
 * @param {object} defaultOptions
 *
 * @returns {function(name: string, description: string, render: function)|function(name: string, description: string, render: function, infoOptions: object)|{ controlled: function(name: string, description: string, render: function, getInitialState: function, infoOptions: object)|function(name: string, description: string, render: function, getInitialState: function, infoOptions: object) }}
 */
export function createStoriesFactory (name, module, defaultOptions = {}) {
  const stories = storiesOf(name, module)

  function addStory (name, description, render, infoOptions = {}) {
    // When rendering function expect any arguments,
    // It means that it's controlled
    if (render.length > 0) {
      render = createController(render)

      infoOptions = {
        ...infoOptions,
        propTablesExclude: [
          render.Controller,
          ...(infoOptions.propTablesExclude || [])
        ]
      }
    }

    // Build options for story
    const options = {
      header: true,
      inline: true,
      text: description,
      ...defaultOptions,
      ...infoOptions
    }

    // Build styles factory
    const stylesFunction = options.styles
      ? stylesheet => options.styles(styles(stylesheet, options), options)
      : stylesheet => styles(stylesheet, options)

    // Create story
    return stories.add(name, withInfo({
      ...options,
      styles: stylesFunction
    })(render))
  }

  return addStory
}

/**
 * Get description from README file
 *
 * @param {string} content
 * @returns {string}
 */
export function getReadmeDescription (content) {
  return content
    .replace(/<p><h1([^>]*)>(.*)<\/h1><\/p>/g, '<h1$1>$2</h1>')
    .replace(/^<h1[^>]*>.*<\/h1>/, '')
    .replace(/(<p>)?<h[1-6][> ][^]*$/i, '')
}
