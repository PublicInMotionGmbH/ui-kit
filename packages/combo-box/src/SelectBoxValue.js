import React from 'react'
import PropTypes from 'prop-types'

import { buildClassName } from '@talixo/shared'

import { Icon } from '@talixo/icon'

import TagsList from './TagsList'

const propTypes = {
  /** Function to build toggle props with handlers */
  getToggleButtonProps: PropTypes.func.isRequired,

  /** Function to render item value */
  renderValue: PropTypes.func.isRequired,

  /** Additional icon to put in the select box */
  icon: PropTypes.node,

  /** Is it multi-select? */
  multi: PropTypes.bool,

  /** Placeholder to show when there is no value */
  placeholder: PropTypes.node,

  /** Is menu open? */
  isOpen: PropTypes.bool,

  /** List of currently selected items */
  selectedItems: PropTypes.array
}

const defaultProps = {
  multi: false,
  isOpen: false,
  selectedItems: []
}

/**
 * Value shown in SelectBox.
 *
 * @param {object} props
 * @param {function} props.getToggleButtonProps
 * @param {function} props.renderValue
 * @param {boolean} props.multi
 * @param {boolean} props.isOpen
 * @param {array} props.selectedItems
 * @param {*} [props.icon]
 * @param {*} [props.placeholder]
 *
 * @returns {React.Element}
 */
function SelectBoxValue (props) {
  const { getToggleButtonProps, icon, multi, tabIndex, placeholder, isOpen, selectedItems, renderValue } = props

  // Get information about selection
  const isSelected = selectedItems.length > 0

  // Get properties required by toggle button
  const passedProps = getToggleButtonProps()

  if (tabIndex != null) {
    passedProps.tabIndex = tabIndex
  }

  // Build class names for elements
  const controlsClsName = buildClassName([ 'combo-box', 'controls' ])
  const innerClsName = buildClassName([ 'combo-box', 'inner' ])
  const arrowClsName = buildClassName([ 'combo-box', 'arrow' ])
  const iconClsName = buildClassName([ 'combo-box', 'icon' ])
  const clsName = buildClassName([ 'combo-box', 'value' ], null, { placeholder: !isSelected })

  // Build icon
  const iconElement = icon ? <div className={iconClsName}>{icon}</div> : null
  const arrow = <Icon name={isOpen ? 'expand_less' : 'expand_more'} />

  // Build controls
  const controls = (
    <div className={controlsClsName}>
      {iconElement}
      <div className={arrowClsName}>{arrow}</div>
    </div>
  )

  // Build component to show value
  const value = multi && isSelected
    ? <TagsList {...props} />
    : isSelected ? renderValue(selectedItems[0], props) : placeholder

  // Render select-box value container
  return (
    <button {...passedProps} className={clsName}>
      <div className={innerClsName}>
        {value}
      </div>
      {controls}
    </button>
  )
}

SelectBoxValue.propTypes = propTypes
SelectBoxValue.defaultProps = defaultProps

export default SelectBoxValue
