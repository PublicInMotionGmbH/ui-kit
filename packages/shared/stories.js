import React from 'react'
import { createStoriesFactory } from './story'

// Load first paragraph from README file
const readme = ''

// Create factories for story
const addStory = createStoriesFactory('Typography and basics', module)

// Stories

addStory('Headers', readme, () => (
  <div>
    <h1>h1. Header</h1>
    <h2>h2. Header</h2>
    <h3>h3. Header</h3>
    <h4>h4. Header</h4>
    <h5>h5. Header</h5>
    <h6>h6. Header</h6>

    <h1>h1. Header <small>Secondary text</small></h1>
    <h2>h2. Header <small>Secondary text</small></h2>
    <h3>h3. Header <small>Secondary text</small></h3>
    <h4>h4. Header <small>Secondary text</small></h4>
    <h5>h5. Header <small>Secondary text</small></h5>
    <h6>h6. Header <small>Secondary text</small></h6>
  </div>
))

addStory('Text styling', readme, () => (
  <div>
    <h2>Paragraphs</h2>

    <p>
      Lorem ipsum dolor sit amet enim. Etiam ullamcorper. Suspendisse a pellentesque dui, non felis.
      Maecenas malesuada elit lectus felis, malesuada ultricies.
      Curabitur et ligula. Ut molestie a, ultricies porta urna.
      Vestibulum commodo volutpat a, convallis ac, laoreet enim. Phasellus fermentum in, dolor.
    </p>

    <p>
      Pellentesque facilisis. Nulla imperdiet sit amet magna.
      Vestibulum dapibus, mauris nec malesuada fames ac turpis velit, rhoncus eu, luctus et interdum adipiscing wisi.
    </p>

    <p>
      Aliquam erat ac ipsum. Integer aliquam purus.
      Quisque lorem tortor fringilla sed, vestibulum id, eleifend justo vel bibendum sapien massa ac turpis
      faucibus orci luctus non, consectetuer lobortis quis, varius in, purus.
      Integer ultrices posuere cubilia Curae, Nulla ipsum dolor lacus, suscipit adipiscing.
      Cum sociis natoque penatibus et ultrices volutpat.
    </p>

    <h2>Inline styling</h2>

    <p>
      We may have some <strong>strong elements</strong> inside our text, as well as some <em>emphasises</em>.
      When we want to calculate 2<sup>3</sup> + x<sub>1</sub>, we may have problems, but it's easy to show it.
      <br />
      In some circumstances, mostly describing our API or other stuff we may also try to show some <code>short code snippet</code>.
      It happens that we may also use <kbd>Ctrl+C</kbd> to hint user with keyboard shortcuts.
    </p>

    <p>
      It may happen that we would like to <mark>highlight</mark> some content inside our dashboards, in case of some search engine.
    </p>

    <p>
      <del>We do not make any mistakes.</del> When we do, we can write it semantically marking that anyway.
    </p>

    <p>
      When we use some shortcuts like <abbr title='Looks good to me'>LGTM</abbr>, we can show such information for user.
    </p>

    <h2>Quotes and testimonials</h2>

    <blockquote>
      This was one of the best experiences I have ever had with a cab company.
      I had problems at the airport and the driver stayed with me for over an hour and helped me sort everything out.
      I would recommend this company to anyone. Thank you for such fabulous service!
    </blockquote>

    <blockquote>
      This was one of the best experiences I have ever had with a cab company.
      I had problems at the airport and the driver stayed with me for over an hour and helped me sort everything out.
      I would recommend this company to anyone. Thank you for such fabulous service!

      <footer>
        <cite>R.M.</cite>
      </footer>
    </blockquote>
  </div>
))

addStory('RTL: Headers', readme, () => (
  <div dir='rtl'>
    <h1>h1. כּוֹתֶרֶת</h1>
    <h2>h2. כּוֹתֶרֶת</h2>
    <h3>h3. כּוֹתֶרֶת</h3>
    <h4>h4. כּוֹתֶרֶת</h4>
    <h5>h5. כּוֹתֶרֶת</h5>
    <h6>h6. כּוֹתֶרֶת</h6>

    <h1>h1. כּוֹתֶרֶת <small>טקסט משני</small></h1>
    <h2>h2. כּוֹתֶרֶת <small>טקסט משני</small></h2>
    <h3>h3. כּוֹתֶרֶת <small>טקסט משני</small></h3>
    <h4>h4. כּוֹתֶרֶת <small>טקסט משני</small></h4>
    <h5>h5. כּוֹתֶרֶת <small>טקסט משני</small></h5>
    <h6>h6. כּוֹתֶרֶת <small>טקסט משני</small></h6>
  </div>
))

addStory('RTL: Text styling', readme, () => (
  <div dir='rtl'>
    <h2>פסקאות</h2>

    <p>
      אחד אל בדפים חרטומים, או פיסול צרפתית ויקיפדיה היא, אם מדריכים הגולשות כתב. ב עוד המחשב ברוכים עקרונות. ספינות מדריכים אקטואליה אחר על.
      זאת אם ביולוגיה אדריכלות, עוד ראשי וקשקש מתמטיקה ב. ויש בה רביעי תרומה לציין. זקוק המדינה גם עוד.
      זאת דת לתרום הבאים נוסחאות, אם פיסיקה סטטיסטיקה אווירונאוטיקה רבה.
    </p>

    <p>
      נפלו למנוע מונחונים או כדי. אחרות תקשורת בה אחר. מיזם יסוד רביעי בה מלא, הבקשה בדפים על זכר. כלל או העזרה ממונרכיה, כלל גם הטבע אנציקלופדיה, אנא אל שמות יסוד.
    </p>

    <p>
תיקונים והגולשים אל בדף, של בדף בישול יוצרים העברית. עזה אם והוא הארץ בידור, תחבורה משופרות עזה אם.
      מה רוסית בארגז הגולשות צ'ט, קבלו מוסיקה ב אחד, מדעי אדריכלות בדף בה. מדע שתפו ויקימדיה אם.
      אדריכלות אווירונאוטיקה אתה של, תרבות ולחבר צ'ט דת.
      לחשבון לויקיפדיה עזה אל, תנך דרכה בכפוף מיוחדים של.
      שתי מה רפואה אודות, כיצד רקטות סטטיסטיקה דת שמו. עוד אקראי וקשקש מיוחדים אם,
      לוח גם העזרה בהתייחסות. מה מונחים אספרנטו קרן. אינו לעריכה של חפש, בה מאמר מוגש אנא.
      בקר הרוח מתמטיקה את, בה החלה יסוד הגולשות קרן. בה לימודים ממונרכיה ואלקטרוניקה עזה,
      חופשית מועמדים מדע של. מדע עזרה אחרות המלצת של, דת יוני אנגלית כתב.
    </p>

    <h2>עיצוב מוטבע</h2>

    <p>
      אולי יש לנו<strong> כמה אלמנטים חזקים בתוך הטקסט </strong>שלנו, כמו גם כמה מדגיש.
      ב במקרים מסוימים, בעיקר לתאר את ה- API שלנו או דברים אחרים אנו עשויים<code> גם לנסות להציג קטע קוד קצר. זה קורה</code>כי אנו יכולים גם >להשתמ
      <kbd>Ctrl+C</kbd> לרמז למשתמש עם קיצורי מקשים.
    </p>

    <p>
      זה יכול לקרות כי אנחנו רוצים להדגיש קצת תוכן בתוך לוחות.<mark> המחוונים שלנו, במקרה של מנוע </mark>החיפוש
    </p>

    <p>
      <del>אנחנו לא עושים שום טעויו</del>ת. כאשר אנו עושים, אנחנו יכולים לכתוב את זה סמנטית סימון כי בכל מקרה.
    </p>

    <p>
      כאשר אנו משתמשים בקיצורי דרך מסוימים כמו <abbr title='נראה לי טוב'>LGTM</abbr>, אנו יכולים להציג מידע כזה עבור המשתמש.
    </p>

    <h2>ציטוטים ועדויות</h2>

    <blockquote>
      זה היה אחד החוויות הטובות ביותר שהיו לי אי פעם עם חברת מוניות.
היו לי בעיות בשדה התעופה והנהג נשאר איתי יותר משעה ועזר לי למיין הכול.
אני ממליץ על חברה זו לכל אחד. תודה לך על שירות נהדר
    </blockquote>

    <blockquote>
      זה היה אחד החוויות הטובות ביותר שהיו לי אי פעם עם חברת מוניות.
היו לי בעיות בשדה התעופה והנהג נשאר איתי יותר משעה ועזר לי למיין הכול.
אני ממליץ על חברה זו לכל אחד. תודה לך על שירות נהדר
      <footer>
        <cite>מה-שמו</cite>
      </footer>
    </blockquote>

    <div style={{ display: 'flex' }}>
      <blockquote style={{ marginLeft: '15px' }}>
        זה היה אחד החוויות הטובות ביותר שהיו לי אי פעם עם חברת מוניות.
  היו לי בעיות בשדה התעופה והנהג נשאר איתי יותר משעה ועזר לי למיין הכול.
  אני ממליץ על חברה זו לכל אחד. תודה לך על שירות נהדר
        <footer>
          <cite>מה-שמו</cite>
        </footer>
      </blockquote>

      <blockquote>
        זה היה אחד החוויות הטובות ביותר שהיו לי אי פעם עם חברת מוניות.
  היו לי בעיות בשדה התעופה והנהג נשאר איתי יותר משעה ועזר לי למיין הכול.
  אני ממליץ על חברה זו לכל אחד. תודה לך על שירות נהדר
        <footer>
          <cite>מה-שמו</cite>
        </footer>
      </blockquote>
    </div>
  </div>
))
