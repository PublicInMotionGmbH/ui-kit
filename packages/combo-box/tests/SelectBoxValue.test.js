import React from 'react'
import { shallow } from 'enzyme'

import { prefix } from '@talixo/shared'

import SelectBoxValue from '../src/SelectBoxValue'

const moduleName = prefix('combo-box', 'value')
const innerName = prefix('combo-box', 'inner')
const iconName = prefix('combo-box', 'icon')
const arrowName = prefix('combo-box', 'arrow')

describe('<SelectBoxValue />', () => {
  it('should render correctly', () => {
    const wrapper = shallow(
      <SelectBoxValue
        getToggleButtonProps={() => {}}
        renderValue={item => item}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render placeholder when there is no single value', () => {
    const wrapper = shallow(
      <SelectBoxValue
        getToggleButtonProps={() => {}}
        renderValue={item => item}
        placeholder={<span id='placeholder' />}
      />
    )

    expect(wrapper.find('#placeholder').length).toBe(1)
  })

  it('should render placeholder when there is no multi value', () => {
    const wrapper = shallow(
      <SelectBoxValue
        multi
        getRemoveButtonProps={() => {}}
        getToggleButtonProps={() => {}}
        renderValue={item => item}
        buildItemId={item => item}
        placeholder={<span id='placeholder' />}
      />
    )

    expect(wrapper.find('#placeholder').length).toBe(1)
  })

  it('should not add information about placeholder to wrapper', () => {
    const wrapper = shallow(
      <SelectBoxValue
        selectedItems={[ 'abc' ]}
        getToggleButtonProps={() => {}}
        renderValue={item => item}
      />
    )

    expect(wrapper.hasClass(`${moduleName}--placeholder`)).toBe(false)
  })

  it('should add information about placeholder to wrapper', () => {
    const wrapper = shallow(
      <SelectBoxValue
        getToggleButtonProps={() => {}}
        renderValue={item => item}
      />
    )

    expect(wrapper.hasClass(`${moduleName}--placeholder`)).toBe(true)
  })

  it('should render single value', () => {
    const wrapper = shallow(
      <SelectBoxValue
        selectedItems={[ 'abc' ]}
        getToggleButtonProps={() => {}}
        renderValue={item => item}
        placeholder={<span id='placeholder' />}
      />
    )

    expect(wrapper.find(`.${innerName}`).text()).toBe('abc')
  })

  it('should render special value', () => {
    const wrapper = shallow(
      <SelectBoxValue
        selectedItems={[ { id: '10' } ]}
        getToggleButtonProps={() => {}}
        renderValue={item => <span id='test'>{item.id}</span>}
        placeholder={<span id='placeholder' />}
      />
    )

    expect(wrapper.find('#test').length).toBe(1)
    expect(wrapper.find('#test').text()).toBe('10')
  })

  it('should not render icon by default', () => {
    const wrapper = shallow(
      <SelectBoxValue
        getToggleButtonProps={() => {}}
        renderValue={item => item}
      />
    )

    expect(wrapper.find(`.${iconName}`).length).toBe(0)
  })

  it('should render provided icon', () => {
    const wrapper = shallow(
      <SelectBoxValue
        icon={<span id='icon' />}
        getToggleButtonProps={() => {}}
        renderValue={item => item}
      />
    )

    expect(wrapper.find(`.${iconName}`).length).toBe(1)
    expect(wrapper.find(`#icon`).length).toBe(1)
  })

  it('should render open icon', () => {
    const wrapper = shallow(
      <SelectBoxValue
        isOpen
        getToggleButtonProps={() => {}}
        renderValue={item => item}
      />
    )

    expect(wrapper.find(`.${arrowName}`).find('Icon').prop('name')).toBe('expand_less')
  })

  it('should render close icon', () => {
    const wrapper = shallow(
      <SelectBoxValue
        isOpen={false}
        getToggleButtonProps={() => {}}
        renderValue={item => item}
      />
    )

    expect(wrapper.find(`.${arrowName}`).find('Icon').prop('name')).toBe('expand_more')
  })

  it('should render list of tags for multi selection with many elements', () => {
    const selectedItems = [ 'abc', 'def' ]
    const getToggleButtonProps = () => {}

    const wrapper = shallow(
      <SelectBoxValue
        multi
        selectedItems={selectedItems}
        isOpen={false}
        getRemoveButtonProps={() => {}}
        getToggleButtonProps={getToggleButtonProps}
        renderValue={item => item}
        buildItemId={item => item}
      />
    )

    expect(wrapper.find('TagsList').length).toBe(1)
    expect(wrapper.find('TagsList').prop('selectedItems')).toBe(selectedItems)
  })
})
