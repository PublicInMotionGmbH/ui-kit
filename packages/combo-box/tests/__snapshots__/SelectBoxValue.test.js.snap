// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectBoxValue /> should render correctly 1`] = `
<button
  className="talixo-combo-box__value talixo-combo-box__value--placeholder"
>
  <div
    className="talixo-combo-box__inner"
  />
  <div
    className="talixo-combo-box__controls"
  >
    <div
      className="talixo-combo-box__arrow"
    >
      <Icon
        name="expand_more"
      />
    </div>
  </div>
</button>
`;
